src/runtime/matlab.cpp

```diff
@@ -0,0 +1,377 @@
+    }
+    // Validate that the dimensionality matches. Matlab is wierd
+    // because matrices always have at least 2 dimensions, and it
+    // truncates trailing dimensions of extent 1. So, the only way
+    // to have an error here is to have more dimensions with
+    // extent != 1 than the Halide pipeline expects.
+    while (dim_count > 0 && mxGetDimensions(arr)[dim_count - 1] == 1) {
+        dim_count--;
+    }
+    if (dim_count > expected_dims) {
+        error(user_context) << "Expected array of rank " << expected_dims
+                            << "for argument " << arg->name
+                            << ", got array of rank " << dim_count << ".\n";
+        return halide_error_code_matlab_bad_param_type;
+    }
+
+    buf->host = (uint8_t *)mxGetData(arr);
+    buf->elem_size = mxGetElementSize(arr);
+
+    // Matlab swaps the first two dimensions, so we need to keep at
+    // least two dimensions for now.
+    if (expected_dims < 2) expected_dims = 2;
+
+    for (int i = 0; i < dim_count && i < expected_dims; i++) {
+        buf->extent[i] = static_cast<int32_t>(mxGetDimensions(arr)[i]);
+    }
+
+    // Add back the dimensions with extent 1.
+    for (int i = 2; i < expected_dims; i++) {
+        if (buf->extent[i] == 0) {
+            buf->extent[i] = 1;
+        }
+    }
+
+    swap(buf->extent[0], buf->extent[1]);
+
+    // Compute dense strides.
+    buf->stride[0] = 1;
+    for (int i = 1; i < expected_dims; i++) {
+        buf->stride[i] = buf->extent[i - 1] * buf->stride[i - 1];
+    }
+
+    return halide_error_code_success;
+}
+
+// Convert a matlab mxArray to a scalar.
+WEAK int halide_mex_array_to_scalar(void *user_context,
+                                    const mxArray *arr, const halide_filter_argument_t *arg, void *scalar) {
+    if (mxIsComplex(arr)) {
+        error(user_context) << "Complex argument not supported for parameter " << arg->name << ".\n";
+        return halide_error_code_generic_error;
+    }
+
+    // Validate that the mxArray has all dimensions of extent 1.
+    int dim_count = mxGetNumberOfDimensions(arr);
+    for (int i = 0; i < dim_count; i++) {
+        if (mxGetDimensions(arr)[i] != 1) {
+            error(user_context) << "Expected scalar argument for parameter " << arg->name << ".\n";
+            return halide_error_code_matlab_bad_param_type;
+        }
+    }
+    if (!mxIsLogical(arr) && !mxIsNumeric(arr)) {
+        error(user_context) << "Expected numeric argument for scalar parameter " << arg->name
+                            << ", got " << mxGetClassName(arr) << ".\n";
+        return halide_error_code_matlab_bad_param_type;
+    }
+
+    double value = mxGetScalar(arr);
+    int32_t type_code = arg->type_code;
+    int32_t type_bits = arg->type_bits;
+
+    if (type_code == halide_type_int) {
+        switch (type_bits) {
+        case 1: *reinterpret_cast<bool *>(scalar) = value != 0; return halide_error_code_success;
```

Comment:
This is one of those rare cases where I'd be OK with a temporary #define to clean up the code.